File: sync/src/packages/source.ts
import fs from "node:fs";
import fsp from "node:fs/promises";
import path from "node:path";

import { rmEntry } from "@runtime/fs.ts";
import { commandExists } from "@runtime/process.ts";
import { runCommand } from "./process.ts";

export { rmEntry } from "@runtime/fs.ts";

const ALPHANUMERIC_PATTERN = /[A-Za-z0-9]/;
const SOURCE_SEPARATOR_PATTERN = /[/:]/;
const TRAILING_PATH_SEPARATOR_PATTERN = /\/+$/;

export function packageCacheDir(cacheRoot: string, source: string): string {
  const slug = sourceSlug(source);
  return path.join(cacheRoot, `${slug}-${fnv1a64(source)}`);
}

export function stagingDirFor(finalDir: string): string {
  const now = BigInt(Date.now()) * 1000000n;
  return withExtension(finalDir, `staging-${process.pid}-${now}`);
}

export async function replaceDirAtomically(src: string, dst: string): Promise<void> {
  const now = BigInt(Date.now()) * 1000000n;
  const backup = withExtension(dst, `backup-${process.pid}-${now}`);
  const legacyBackup = withExtension(dst, "backup");
  if (exists(legacyBackup)) {
    rmEntry(legacyBackup);
  }
  rmEntry(backup);
  let movedToBackup = false;
  if (exists(dst)) {
    await fsp.rename(dst, backup);
    movedToBackup = true;
  }

  try {
    await fsp.rename(src, dst);
    if (movedToBackup) {
      rmEntry(backup);
    }
  } catch (error) {
    if (movedToBackup && exists(backup)) {
      try {
        await fsp.rename(backup, dst);
      } catch {
        // best-effort rollback only
      }
    }
    throw error;
  }
}

export async function clonePackage(
  source: string,
  targetDir: string,
  timeoutMs: number,
): Promise<boolean> {
  // clonePackageWithRunner removes the target directory between attempts and after
  // every failure. We only call it with a private staging directory from stagingDirFor.
  return clonePackageWithRunner(source, targetDir, await commandExists("gh"), (command) =>
    runCommand(command, undefined, timeoutMs, "clone"),
  );
}

function sourceSlug(source: string): string {
  const trimmed = source.trim().replace(TRAILING_PATH_SEPARATOR_PATTERN, "");
  const normalized = trimmed.endsWith(".git") ? trimmed.slice(0, -4) : trimmed;
  const sourceParts = isLocalPathSource(normalized)
    ? [localPathBasename(normalized)]
    : normalized
        .split(SOURCE_SEPARATOR_PATTERN)
        .filter((part) => part.length > 0)
        .slice(-2);
  const joined = sourceParts.length === 0 ? "package" : sourceParts.join("-");
  const sanitized = joined
    .split("")
    .map((ch) => (ALPHANUMERIC_PATTERN.test(ch) ? ch.toLowerCase() : "-"))
    .join("");
  const compact = sanitized
    .split("-")
    .filter((part) => part.length > 0)
    .join("-");
  return compact.length > 0 ? compact : "package";
}

const localPathBasename = (source: string): string => path.basename(source);

const isLocalPathSource = (source: string): boolean => path.isAbsolute(source);

function fnv1a64(input: string): string {
  let hash = 0xcbf29ce484222325n;
  for (const byte of new TextEncoder().encode(input)) {
    hash ^= BigInt(byte);
    hash = (hash * 0x100000001b3n) & 0xffffffffffffffffn;
  }
  return hash.toString(16).padStart(16, "0");
}

function withExtension(target: string, extension: string): string {
  const dir = path.dirname(target);
  const base = path.basename(target);
  const ext = path.extname(base);
  const stem = ext ? base.slice(0, -ext.length) : base;
  return path.join(dir, `${stem}.${extension}`);
}

export async function clonePackageWithRunner(
  source: string,
  targetDir: string,
  ghAvailable: boolean,
  runner: (command: readonly string[]) => Promise<boolean>,
): Promise<boolean> {
  const commands = cloneCommands(source, targetDir, ghAvailable);
  let index = 0;
  for (const command of commands) {
    if (index > 0) {
      rmEntry(targetDir);
    }
    index += 1;
    if (await runner(command)) {
      return true;
    }
    rmEntry(targetDir);
  }
  return false;
}

function cloneCommands(source: string, targetDir: string, ghAvailable: boolean): string[][] {
  const commands: string[][] = [];
  const slug = githubRepoSlug(source);
  if (slug && ghAvailable) {
    commands.push(["gh", "repo", "clone", slug, targetDir, "--", "--depth=1"]);
  }
  commands.push(["git", "clone", "--depth=1", source, targetDir]);
  return commands;
}

function githubRepoSlug(source: string): string | null {
  const trimmed = source.trim();
  const normalized = trimmed.endsWith(".git") ? trimmed.slice(0, -4) : trimmed;
  if (normalized.startsWith("https://github.com/")) {
    return splitOwnerRepo(normalized.slice("https://github.com/".length));
  }
  if (normalized.startsWith("http://github.com/")) {
    return splitOwnerRepo(normalized.slice("http://github.com/".length));
  }
  if (normalized.startsWith("git@github.com:")) {
    return splitOwnerRepo(normalized.slice("git@github.com:".length));
  }
  return null;
}

function splitOwnerRepo(rest: string): string | null {
  const parts = rest
    .split("/")
    .filter((part) => part.length > 0)
    .slice(0, 2);
  if (parts.length !== 2) {
    return null;
  }
  return `${parts[0]}/${parts[1]}`;
}

function exists(target: string): boolean {
  try {
    fs.accessSync(target);
    return true;
  } catch {
    return false;
  }
}

File: sync/src/sync/packages/source.py
# Copyright (c) 2026 agents-sync. SPDX-License-Identifier: AGPL-3.0-or-later
"""Package source resolution, FNV-1a64 hashing, atomic operations, and Git cloning."""

from __future__ import annotations

import asyncio
import contextlib
import os
import re
import time
from pathlib import Path
from typing import TYPE_CHECKING

from sync.runtime.fs import rm_entry
from sync.runtime.process import command_exists, run_command

if TYPE_CHECKING:
    from collections.abc import Awaitable, Callable, Sequence

__all__ = [
    "clone_package",
    "clone_package_with_runner",
    "fnv1a64",
    "package_cache_dir",
    "replace_dir_atomically",
    "rm_entry",
    "source_slug",
    "staging_dir_for",
]

_ALPHANUMERIC_PATTERN = re.compile(r"[A-Za-z0-9]")
_SOURCE_SEPARATOR_PATTERN = re.compile(r"[/:]")
_TRAILING_PATH_SEPARATOR_PATTERN = re.compile(r"/+$")
_FNV_OFFSET_BASIS = 0xCBF29CE484222325
_FNV_PRIME = 0x100000001B3
_MASK_64 = 0xFFFFFFFFFFFFFFFF
_OWNER_REPO_PARTS_COUNT = 2


def package_cache_dir(cache_root: str, source: str) -> str:
    """Compute local cache directory path for a given package source."""
    slug = source_slug(source)
    return str(Path(cache_root) / f"{slug}-{fnv1a64(source)}")


def staging_dir_for(final_dir: str) -> str:
    """Generate a unique staging directory path for atomic package installation."""
    now = time.time_ns()
    return _with_extension(final_dir, f"staging-{os.getpid()}-{now}")


def _replace_dir_atomically_sync(src: str, dst: str) -> None:
    now = time.time_ns()
    backup = _with_extension(dst, f"backup-{os.getpid()}-{now}")
    legacy_backup = _with_extension(dst, "backup")

    if _exists(legacy_backup):
        rm_entry(legacy_backup)
    rm_entry(backup)

    moved_to_backup = False
    if _exists(dst):
        Path(dst).replace(backup)
        moved_to_backup = True

    try:
        Path(src).replace(dst)
        if moved_to_backup:
            rm_entry(backup)
    except Exception:
        if moved_to_backup and _exists(backup):
            with contextlib.suppress(OSError):
                Path(backup).replace(dst)
        raise


async def replace_dir_atomically(src: str, dst: str) -> None:
    """Atomically replace dst with src directory, rolling back on failure."""
    await asyncio.to_thread(_replace_dir_atomically_sync, src, dst)


async def clone_package(
    source: str,
    target_dir: str,
    timeout_ms: int,
) -> bool:
    """Clone git repository package into target directory using gh or git."""
    gh_available = await command_exists("gh")

    async def runner(command: Sequence[str]) -> bool:
        return await run_command(
            command, cwd=None, timeout_ms=timeout_ms, action="clone"
        )

    return await clone_package_with_runner(
        source, target_dir, gh_available=gh_available, runner=runner
    )


def source_slug(source: str) -> str:
    """Generate a clean alphanumeric slug from a package source URL or path."""
    trimmed = _TRAILING_PATH_SEPARATOR_PATTERN.sub("", source.strip())
    normalized = trimmed.removesuffix(".git")
    if _is_local_path_source(normalized):
        source_parts = [_local_path_basename(normalized)]
    else:
        parts = [p for p in _SOURCE_SEPARATOR_PATTERN.split(normalized) if p]
        source_parts = parts[-2:]

    joined = "-".join(source_parts) if source_parts else "package"
    sanitized = "".join(
        ch.lower() if _ALPHANUMERIC_PATTERN.match(ch) else "-" for ch in joined
    )
    compact = "-".join(part for part in sanitized.split("-") if part)
    return compact or "package"


def fnv1a64(input_str: str) -> str:
    """Compute 64-bit FNV-1a hash formatted as 16 lowercase hex digits."""
    hash_val = _FNV_OFFSET_BASIS
    for byte in input_str.encode("utf-8"):
        hash_val ^= byte
        hash_val = (hash_val * _FNV_PRIME) & _MASK_64
    return f"{hash_val:016x}"


def _local_path_basename(source: str) -> str:
    return Path(source).name


def _is_local_path_source(source: str) -> bool:
    return Path(source).is_absolute()


def _with_extension(target: str, extension: str) -> str:
    target_path = Path(target)
    return str(target_path.parent / f"{target_path.stem}.{extension}")


async def clone_package_with_runner(
    source: str,
    target_dir: str,
    *,
    gh_available: bool,
    runner: Callable[[Sequence[str]], Awaitable[bool]],
) -> bool:
    """Attempt clone commands sequentially, removing target on failure/retry."""
    commands = _clone_commands(source, target_dir, gh_available=gh_available)
    for index, command in enumerate(commands):
        if index > 0:
            rm_entry(target_dir)
        if await runner(command):
            return True
        rm_entry(target_dir)
    return False


def _clone_commands(
    source: str,
    target_dir: str,
    *,
    gh_available: bool,
) -> list[list[str]]:
    commands: list[list[str]] = []
    slug = _github_repo_slug(source)
    if slug and gh_available:
        commands.append(["gh", "repo", "clone", slug, target_dir, "--", "--depth=1"])
    commands.append(["git", "clone", "--depth=1", source, target_dir])
    return commands


def _github_repo_slug(source: str) -> str | None:
    trimmed = source.strip()
    normalized = trimmed.removesuffix(".git")
    if normalized.startswith("https://github.com/"):
        return _split_owner_repo(normalized.removeprefix("https://github.com/"))
    if normalized.startswith("http://github.com/"):
        return _split_owner_repo(normalized.removeprefix("http://github.com/"))
    if normalized.startswith("git@github.com:"):
        return _split_owner_repo(normalized.removeprefix("git@github.com:"))
    return None


def _split_owner_repo(rest: str) -> str | None:
    parts = [part for part in rest.split("/") if part][:_OWNER_REPO_PARTS_COUNT]
    if len(parts) != _OWNER_REPO_PARTS_COUNT:
        return None
    return f"{parts[0]}/{parts[1]}"


def _exists(target: str) -> bool:
    try:
        return Path(target).exists()
    except OSError:
        return False
